File: frontend/reddit-frontend-starter/src/services/postService.js
import api from "./api";

/* =========================================
   Get All Posts
========================================= */

export const getPosts = async () => {

  try {

    const response =
      await api.get(
        "/api/posts"
      );

    return response.data;

  } catch (error) {

    console.log(
      "Get Posts Error :",
      error.response?.data || error
    );

    return [];

  }

};

/* =========================================
   Create Post
========================================= */

export const createPost =
  async (data) => {

    try {

      const response =
        await api.post(

          "/api/posts",

          data

        );

      return response.data;

    } catch (error) {

      console.log(
        "Create Post Error :",
        error.response?.data || error
      );

      throw error;

    }

  };

File: backend/reddit-backend-starter/src/main/java/com/socialmedia/backend/service/PostService.java
package com.socialmedia.backend.service;

import com.socialmedia.backend.dto.request.CreatePostRequest;
import com.socialmedia.backend.dto.response.PostResponse;

import java.util.List;

public interface PostService {

    /* =========================================
       Create New Post
    ========================================= */

    PostResponse createPost(
            CreatePostRequest request
    );

    /* =========================================
       Get All Posts
    ========================================= */

    List<PostResponse> getAllPosts();

    /* =========================================
       Search Posts
    ========================================= */

    List<PostResponse> searchPosts(
            String keyword
    );

    /* =========================================
       Get Post By Id
    ========================================= */

    PostResponse getPostById(
            Long id
    );

    /* =========================================
       Get Posts By Community
    ========================================= */

    List<PostResponse> getPostsByCommunity(
            Long communityId
    );

    /* =========================================
       Update Post
    ========================================= */

    PostResponse updatePost(
            Long id,
            CreatePostRequest request
    );

    /* =========================================
       Upvote Post
    ========================================= */

    PostResponse upvotePost(
            Long id
    );

    /* =========================================
       Downvote Post
    ========================================= */

    PostResponse downvotePost(
            Long id
    );

    /* =========================================
       Delete Post
    ========================================= */

    void deletePost(
            Long id
    );

}